memorystore/valkey/leaderboard/app/src/main/java/app/DataController.java

```diff
@@ -0,0 +1,170 @@
+   */
+  public DataController(final Jedis redisClient,
+      final LeaderboardRepository repository) {
+    this.leaderboardRepository = repository;
+    this.jedis = redisClient;
+  }
+
+  /**
+   * Get the leaderboard entries starting from the given position.
+   *
+   * @param position The starting position of the entries to search.
+   * @param orderBy  The order of the entries.
+   * @param pageSize The number of entries to return.
+   * @param username The username to check the rank of.
+   * @return The leaderboard entries.
+   */
+  public LeaderboardResponse getLeaderboard(
+      final long position,
+      final OrderByType orderBy,
+      final long pageSize,
+      final String username) {
+    String cacheKey = Global.LEADERBOARD_ENTRIES_KEY;
+    long maxPosition = position + pageSize - 1;
+
+    // Initialize the cache if it's empty
+    boolean cacheUpdated = this.initializeCache();
+
+    // Set the cache status for the front end
+    int cacheStatus = cacheUpdated
+        ? FromCacheType.FROM_DB.getValue()
+        : FromCacheType.FULL_CACHE.getValue();
+
+    // If we have a username, search for the user's rank
+    if (username != null) {
+      Long userRank = jedis.zrevrank(cacheKey, username);
+      if (userRank != null) {
+        long pos = userRank;
+        long maxPos = userRank + pageSize - 1;
+
+        return new LeaderboardResponse(
+            getEntries(cacheKey, pos, maxPos, true),
+            cacheStatus);
+      }
+    }
+
+    // Get the leaderboard entries depending on the order
+    List<LeaderboardEntry> leaderboardList = getEntries(
+        cacheKey, position, maxPosition,
+        orderBy == OrderByType.HIGH_TO_LOW);
+
+    return new LeaderboardResponse(leaderboardList, cacheStatus);
+  }
+
+  private List<LeaderboardEntry> getEntries(
+      final String cacheKey,
+      final long position,
+      final long maxPosition,
+      final boolean isDescending) {
+    // Define an object
+    List<Tuple> entries = new ArrayList<>();
+
+    // Use zrevrangeWithScores to get the entries in descending order
+    if (isDescending) {
+      entries = new ArrayList<>(
+          jedis.zrevrangeWithScores(cacheKey, position, maxPosition));
+    }
+
+    // If zrangeWithScores is used, the entries are in ascending order
+    if (!isDescending) {
+      entries = new ArrayList<>(
+          jedis.zrangeWithScores(cacheKey, position, maxPosition));
+    }
+
+    List<LeaderboardEntry> newEntries = new ArrayList<>();
+    for (int i = 0; i < entries.size(); i++) {
+      Tuple e = entries.get(i);
+
+      // Calculate overall position
+      long overallPosition = position + i;
+      if (!isDescending) {
+        overallPosition = jedis.zcard(cacheKey) - overallPosition - 1;
+      }
+
+      newEntries.add(
+          new LeaderboardEntry(
+            e.getElement(), e.getScore(), overallPosition
+          )
+      );
+    }
+
+    return newEntries;
+  }
+
+  private boolean initializeCache() {
```

Comment:
document what the return value means? Looks like it is true if an initialization had to happen?